### ISSUE TITLE
Modify loss function to consider class imbalance

### ISSUE THREAD

**Issue by kaustubhharapanahalli** [AlexeyAB/darknet#4792]
Hi, 

Where can I modify the loss function, so that I can take care of class imbalance?

Thanks
Kaustubh

**Comment by AlexeyAB** [comment]
Do either
```
[yolo]
focal_loss=1
```

or
```
[yolo]
counters_per_class=100, 2000, 300, ... # number of objects per class in your Training dataset
```

**Comment by kaustubhharapanahalli** [comment]
oh okay, I'll try both of them. Thanks

**Comment by Gaondong** [comment]
> Do either
> 
> ```
> [yolo]
> focal_loss=1
> ```
> 
> or
> 
> ```
> [yolo]
> counters_per_class=100, 2000, 300, ... # number of objects per class in your Training dataset
> ```

Sorry to bother, what does the **counters_per_class** do?  
Is it limit the number of objects of whole dataset? 
For example,
My dataset have two classes(car,person). In **8000 train dataset**,  car has **50000 bbox**, person only has **4000 bbox**,.
How do I set this parameter？

**Comment by kaustubhharapanahalli** [comment]
@Gaondong 
counters_per_class is the number of objects that are present for each class in your dataset. This number depends on the number of objects that you are pushing for training. If you are using 100% of your dataset, then it should have the total count, else based on the number of images that you are using for training.

For the example you have given, if your car id is 1 and person is 2 then,
`counters_per_class=50000, 4000`

**Comment by Gaondong** [comment]
> @Gaondong
> counters_per_class is the number of objects that are present for each class in your dataset. This number depends on the number of objects that you are pushing for training. If you are using 100% of your dataset, then it should have the total count, else based on the number of images that you are using for training.
> 
> For the example you have given, if your car id is 1 and person is 2 then,
> `counters_per_class=50000, 4000`

Did it work for your dataset? 
How does **counters_per_class** affect loss function？

**Comment by AlexeyAB** [comment]
loss_class_1 = 4000/(50000+4000)
loss_class_2 = 50000/(50000+4000)

**Comment by Gaondong** [comment]
> loss_class_1 = 4000/(50000+4000)
> loss_class_2 = 50000/(50000+4000)

Thanks for your reply.
Can focal loss solve this problem?

**Comment by Gaondong** [comment]
> loss_class_1 = 4000/(50000+4000)
> loss_class_2 = 50000/(50000+4000)

@AlexeyAB 
Sorry to bother you, I really hope to get your help.
For imbalance in every picture of train dataset: In **8000 train dataset**, **car has 50000 bbox**, **person only has 4000 bbox**,.
Which is more effective to modify **focal_loss** or **counters_per_class** in your experience? 
Or both effects are small？

**Comment by AlexeyAB** [comment]
counters_per_class= is more effective.

But you can try 3 options:
* focal_loss 
* counters_per_class 
* focal_loss  + counters_per_class

**Comment by Gaondong** [comment]
> counters_per_class= is more effective.
> 
> But you can try 3 options:
> 
> * focal_loss
> * counters_per_class
> * focal_loss  + counters_per_class

Thank you very much I will try these three ways！

**Comment by Gaondong** [comment]
> counters_per_class= is more effective.
> 
> But you can try 3 options:
> 
> * focal_loss
> * counters_per_class
> * focal_loss  + counters_per_class

counters_per_class is more effective +3%map, but loss is still higher than 4 at 30,000 iterations with map not increase.
focal loss reduce map, 
facal_loss+counter_per_class help less.

**Comment by AlexeyAB** [comment]
@Gaondong 

I added commit: https://github.com/AlexeyAB/darknet/commit/3d9aa2af4718a3bd5bbe23de2022987cb767e9c5
Now you can use

`./darknet detector calc_anchors data/coco.data -num_of_clusters 9 -width 416 -height 416`
to calculate both
* `anchors=` will be saved to the `anchors.txt` file
* `counters_per_class=` will be saved to the `counters_per_class.txt` file

**Comment by AlexeyAB** [comment]
@WongKinYiu Did you try to train yolov3-tiny-prn with `counters_per_class=`  for each [yolo] layer to solve class imbalance?

For MS COCO:
```
[yolo]
counters_per_class = 265751, 7113, 44338, 8730, 5154, 6100, 4550, 10038, 10832, 13016, 1883, 1974, 1287, 9787, 10883, 4776, 5507, 6562, 9724, 8171, 5491, 1312, 5296, 5193, 8707, 11531, 12415, 6470, 6214, 2688, 6605, 2665, 6393, 9159, 3300, 3756, 5508, 6143, 4854, 24516, 7914, 20710, 5466, 7810, 6168, 14398, 9603, 5941, 4403, 6561, 7308, 7885, 2908, 5892, 7401, 6466, 38878, 5804, 8585, 4160, 15789, 4157, 5836, 4964, 2273, 5746, 2891, 6411, 1641, 3335, 223, 5624, 2658, 25244, 6320, 6539, 1464, 4775, 198, 1937

```

**Comment by WongKinYiu** [comment]
@AlexeyAB 

I got buffer overflow error after training 770 iterations.

I set counter_per_class of every yolo layers, do i only set it for last yolo layer?

**Comment by WongKinYiu** [comment]
@AlexeyAB Hello,

I have checked the code, it is because `(l.classes_multipliers) ?` in following three lines.
https://github.com/AlexeyAB/darknet/blob/master/src/yolo_layer.c#L427
https://github.com/AlexeyAB/darknet/blob/master/src/yolo_layer.c#L469
https://github.com/AlexeyAB/darknet/blob/master/src/yolo_layer.c#L516
after fixed these three lines, no `buffer overflow detected` occurs.

**Comment by Gaondong** [comment]
> counter_per_class

I set counter_per_class for every yolo layer in cfg document and did not get any error.

**Comment by Gaondong** [comment]
@WongKinYiu 
I set counter_per_class for every yolo layer in cfg document and did not get any error.

**Comment by WongKinYiu** [comment]
@Gaondong Thanks,

Hmm.. yes, it seems be other problems.
my training get same error after several thousands iterations.
i will keep debugging...

**Comment by spaul13** [comment]
@WongKinYiu @AlexeyAB @cenit @kaustubhharapanahalli @Gaondong, I have a classification model that doesn't contain the yolo layer. Can anyone please tell me how to solve the class imbalance problem for a classification model?

**Comment by zpmmehrdad** [comment]
@AlexeyAB  Hi,

I have trained a model with counter_per_class but the loss is high and after almost 1.5K interactions I get avg nan. I tested it without counter_per_class and the loss is normal and everything is fine. Whats is the problem? I have 58 classes and this is my counter_per_class:

`counters_per_class = 13859, 755, 199, 2109, 839, 766, 2580, 709, 278, 633, 86, 346, 28, 245, 2342, 548, 1144, 49, 4117, 1622, 263, 251, 2484, 1007, 176, 671, 157, 18, 31, 2698, 11954, 2692, 8982, 2907, 495, 1106, 5192, 2382, 4313, 101, 6941, 16945, 26027, 3781, 1426, 2200, 1070, 272, 197, 124, 2203, 1967, 707, 118, 260, 127, 38, 924`

Thanks

**Comment by AmanGoyal99** [comment]
Both focal_loss and counters_per_class are to be manually set by us in the cfg document because there are no such predefined parameters as such.
Also is both of the parameters mentioned above is to be set for all the yolo layers?

**Comment by arnaud-nt2i** [comment]
@AlexeyAB, @WongKinYiu  Hi, !
With counters_per_class I got a very high avg loss (around 150) after( 7000 iter) and a lot of amplitude in loss variation.
I am using a big dataset (more than 200000 pics) and training is still near from start but in the same conditions, without counters_per_class  loss is already nearby 5... (yolo sam mish model)
Do I need to persist? (we don't care about loss)
Or is something wrong with counters_per_class ?

**Comment by htdung167** [comment]
@AlexeyAB  i saw "Unused field: 'counters_per_class = 2875,6145,3900,3105,3465,6235,3270,3210,2690,..." when i start training. what happened?

**Comment by stephanecharette** [comment]
Did you perhaps set it in the wrong section?  Show us the 5 lines above and 5 lines below where you set it.

**Comment by stephanecharette** [comment]
YOLOv2 hasn't been used in many years.  Does it even have support for `counter_per_class`?

Note that either way, you've put it in the wrong section.  It should be in the [yolo] section.

**Comment by htdung167** [comment]
> YOLOv2 hasn't been used in many years. Does it even have support for `counters_per_class`?
> 
> Note that either way, you've put it in the wrong section. It should be in the [yolo] section.

oh, thanks a lot.
